File: java/src/main/java/trinsic/services/CredentialTemplateService.java
package trinsic.services;

import com.google.common.util.concurrent.ListenableFuture;
import com.google.protobuf.InvalidProtocolBufferException;
import trinsic.okapi.DidException;
import trinsic.sdk.options.v1.Options;
import trinsic.services.verifiablecredentials.templates.v1.*;
import trinsic.services.verifiablecredentials.templates.v1.CredentialTemplatesGrpc;

public class CredentialTemplateService extends ServiceBase {
  private final CredentialTemplatesGrpc.CredentialTemplatesFutureStub stub;

  public CredentialTemplateService() {
    this(null);
  }

  public CredentialTemplateService(Options.ServiceOptions.Builder options) {
    super(options);

    this.stub = CredentialTemplatesGrpc.newFutureStub(this.getChannel());
  }

  public ListenableFuture<CreateCredentialTemplateResponse> create(
      CreateCredentialTemplateRequest request) throws InvalidProtocolBufferException, DidException {
    return withMetadata(stub, request).create(request);
  }

  public ListenableFuture<GetCredentialTemplateResponse> get(GetCredentialTemplateRequest request)
      throws InvalidProtocolBufferException, DidException {
    return withMetadata(stub, request).get(request);
  }

  public ListenableFuture<ListCredentialTemplatesResponse> list(
      ListCredentialTemplatesRequest request) throws InvalidProtocolBufferException, DidException {
    return withMetadata(stub, request).list(request);
  }

  public ListenableFuture<SearchCredentialTemplatesResponse> search(
      SearchCredentialTemplatesRequest request)
      throws InvalidProtocolBufferException, DidException {
    return withMetadata(stub, request).search(request);
  }

  public ListenableFuture<DeleteCredentialTemplateResponse> delete(
      DeleteCredentialTemplateRequest request) throws InvalidProtocolBufferException, DidException {
    return withMetadata(stub, request).delete(request);
  }
}

File: go/services/credentialtemplate_service.go
package services

import (
	"context"

	"github.com/trinsic-id/sdk/go/proto/services/verifiablecredentials/templates/v1/template"
)

// NewCredentialTemplateService returns a credential template servcie with the base service configured
// using the provided options
func NewCredentialTemplateService(options *Options) (CredentialTemplateService, error) {
	base, err := NewServiceBase(options)
	if err != nil {
		return nil, err
	}
	service := &credentialTemplateBase{
		Service: base,
		client:  template.NewCredentialTemplatesClient(base.GetChannel()),
	}

	return service, nil
}

// CredentialTemplateService defines the interface for interacting with credentials templates
type CredentialTemplateService interface {
	Service
	// Create a credential template
	Create(userContext context.Context, request *template.CreateCredentialTemplateRequest) (*template.CreateCredentialTemplateResponse, error)
	// Get a specific credential template
	Get(userContext context.Context, request *template.GetCredentialTemplateRequest) (*template.GetCredentialTemplateResponse, error)
	// List available credential templates
	List(userContext context.Context, request *template.ListCredentialTemplatesRequest) (*template.ListCredentialTemplatesResponse, error)
	// Search for a template
	Search(userContext context.Context, request *template.SearchCredentialTemplatesRequest) (*template.SearchCredentialTemplatesResponse, error)
	// Delete a template
	Delete(userContext context.Context, request *template.DeleteCredentialTemplateRequest) (*template.DeleteCredentialTemplateResponse, error)
}

type credentialTemplateBase struct {
	Service
	client template.CredentialTemplatesClient
}

func (c *credentialTemplateBase) Create(userContext context.Context, request *template.CreateCredentialTemplateRequest) (*template.CreateCredentialTemplateResponse, error) {
	md, err := c.GetMetadataContext(userContext, request)
	if err != nil {
		return nil, err
	}
	response, err := c.client.Create(md, request)
	if err != nil {
		return nil, err
	}
	return response, nil
}

func (c *credentialTemplateBase) Get(userContext context.Context, request *template.GetCredentialTemplateRequest) (*template.GetCredentialTemplateResponse, error) {
	md, err := c.GetMetadataContext(userContext, request)
	if err != nil {
		return nil, err
	}
	response, err := c.client.Get(md, request)
	if err != nil {
		return nil, err
	}
	return response, nil
}

func (c *credentialTemplateBase) List(userContext context.Context, request *template.ListCredentialTemplatesRequest) (*template.ListCredentialTemplatesResponse, error) {
	md, err := c.GetMetadataContext(userContext, request)
	if err != nil {
		return nil, err
	}
	response, err := c.client.List(md, request)
	if err != nil {
		return nil, err
	}
	return response, nil
}

func (c *credentialTemplateBase) Search(userContext context.Context, request *template.SearchCredentialTemplatesRequest) (*template.SearchCredentialTemplatesResponse, error) {
	md, err := c.GetMetadataContext(userContext, request)
	if err != nil {
		return nil, err
	}
	response, err := c.client.Search(md, request)
	if err != nil {
		return nil, err
	}
	return response, nil
}

func (c *credentialTemplateBase) Delete(userContext context.Context, request *template.DeleteCredentialTemplateRequest) (*template.DeleteCredentialTemplateResponse, error) {
	md, err := c.GetMetadataContext(userContext, request)
	if err != nil {
		return nil, err
	}
	response, err := c.client.Delete(md, request)
	if err != nil {
		return nil, err
	}
	return response, nil
}
